src/interrupt/mod.rs

```diff
@@ -0,0 +1,142 @@
+pub type Result<T> = std::io::Result<T>;
+
+/// Data type to store an interrupt source identifier.
+pub type InterruptIndex = u32;
+
+/// Data type to store an interrupt source type.
+pub type InterruptType = u32;
+
+pub const PIN_IRQ: InterruptType = 0;
+pub const PCI_MSI_IRQ: InterruptType = 1;
+
+/// Data type to store an interrupt event.
+pub type InterruptEvent = u32;
+
+pub const IRQ_TRIGGERED: InterruptEvent = 0;
+
+/// Trait to manage interrupt sources for virtual device backends.
+///
+/// The InterruptManager implementations should protect itself from concurrent accesses internally,
+/// so it could be invoked from multi-threaded context.
+pub trait InterruptManager {
+    /// Create an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object to manage
+    /// interrupt sources for a virtual device
+    ///
+    /// An [InterruptSourceGroup](trait.InterruptSourceGroup.html) object manages all interrupt
+    /// sources of the same type for a virtual device.
+    ///
+    /// # Arguments
+    /// * interrupt_type: type of interrupt source.
+    /// * base: base Interrupt Source ID to be managed by the group object.
+    /// * count: number of Interrupt Sources to be managed by the group object.
+    fn create_group(
+        &self,
+        interrupt_type: InterruptType,
+        base: InterruptIndex,
+        count: InterruptIndex,
+    ) -> Result<Arc<Box<dyn InterruptSourceGroup>>>;
+
+    /// Destroy an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object created by
+    /// [create_group()](trait.InterruptManager.html#tymethod.create_group).
+    ///
+    /// Assume the caller takes the responsibility to disable all interrupt sources of the group
+    /// before calling destroy_group(). This assumption helps to simplify InterruptSourceGroup
+    /// implementations.
+    fn destroy_group(&self, group: Arc<Box<dyn InterruptSourceGroup>>) -> Result<()>;
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct MsiIrqSourceConfig {
+    /// High address to delivery message signaled interrupt.
+    pub high_addr: u32,
+    /// Low address to delivery message signaled interrupt.
+    pub low_addr: u32,
+    /// Data to write to delivery message signaled interrupt.
+    pub data: u32,
+}
+
+pub trait InterruptSourceGroup: Send + Sync {
+    /// Enable the interrupt sources in the group to generate interrupts.
+    fn enable(&self) -> Result<()>;
+
+    /// Disable the interrupt sources in the group to generate interrupts.
+    fn disable(&self) -> Result<()>;
+
+    /// Inject an interrupt from this interrupt source into the guest.
+    fn trigger(&self, index: InterruptIndex) -> Result<()>;
+
+    /// Returns an interrupt notifier from this interrupt.
+    /// An interrupt notifier allows for external components and processes
+    /// to inject interrupts into a guest, by writing to the file returned
+    /// by this method.
+    fn notifier(&self, index: InterruptIndex) -> Option<File>;
```

Comment:
Not sure if File is the best option here. Maybe we can have something that implement the traits needed (write, read?). I don't know which traits are needed actually. I'll experiment with this a bit.

Reply:
We should also have a default implementation that returns none as this should not be implemented by non vhost-user backends.

Reply:
> Not sure if File is the best option here. Maybe we can have something that implement the traits needed (write, read?). I don't know which traits are needed actually. I'll experiment with this a bit.

It's not really about traits that need to be implemented but about the fact that it can be made into a raw fd, to pass it down to the vhost-user backend process.
A `File` is something that would work (That's the inner type for `Eventfd`...) and is the least common denominator I found between Windows and Linux.